tremor-script/src/registry/custom_aggregate_fn.rs

```diff
@@ -0,0 +1,237 @@
+use crate::registry::{FResult, FunctionError};
+use crate::{AggrType, Value};
+use beef::Cow;
+
+/// public because lalrpop
+#[derive(Clone)]
+pub struct CustomAggregateFn<'script> {
+    /// public because lalrpop
+    pub name: Cow<'script, str>,
+    /// public because lalrpop
+    pub init_body: Exprs<'script>,
+    /// public because lalrpop
+    pub aggregate_args: Vec<String>,
+    /// public because lalrpop
+    pub aggregate_body: Exprs<'script>,
+    /// public because lalrpop
+    pub mergein_args: Vec<String>,
+    /// public because lalrpop
+    pub mergein_body: Exprs<'script>,
+    /// public because lalrpop
+    pub emit_args: Vec<String>,
+    /// public because lalrpop
+    pub emit_body: Exprs<'script>,
+    /// public because lalrpop
+    pub state: Value<'script>,
+}
+
+impl<'script> CustomAggregateFn<'script> {
+    /// Initialize the instance of aggregate function
+    /// # Errors
+    /// Returns an error if the initialization fails, or the init function does not return a value
+    pub fn init<'event>(&mut self, env: &Env<'_, 'event>) -> FResult<()>
+    where
+        'script: 'event,
+    {
+        let mut local_stack = LocalStack::with_size(128);
+
+        let mut no_meta = Value::null();
+        let mut state = Value::null().into_static();
+        let mut no_event = Value::null();
+
+        let mut body_iter = self.init_body.iter().peekable();
+        while let Some(expr) = body_iter.next() {
+            let env_local = Env {
+                context: env.context,
+                consts: env.consts,
+                aggrs: &[],
+                meta: env.meta,
+                recursion_limit: env.recursion_limit,
+            };
+
+            let cont = expr.run(
+                ExecOpts {
+                    result_needed: true,
+                    aggr: AggrType::Tick,
+                },
+                &env_local,
+                &mut no_event,
+                &mut state,
+                &mut no_meta,
+                &mut local_stack,
+            )?;
+
+            if body_iter.peek().is_none() {
+                if let Cont::Cont(value) = cont {
+                    self.state = value.into_owned().clone_static();
+                } else {
+                    return Err(FunctionError::NoAggregateValueInInit);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Aggregate a value
+    /// # Errors
+    /// Will return an error if the aggregate function does not return a value, of it errors
+    pub fn aggregate<'event>(&mut self, args: &[&Value], env: &Env<'_, 'event>) -> FResult<()>
+    where
+        'script: 'event,
+    {
+        let mut body_iter = self.aggregate_body.iter().peekable();
+        let mut no_meta = Value::null();
+        let mut state = Value::null().into_static();
+        let mut no_event = Value::null();
+        let mut local_stack = LocalStack::with_size(2);
+        local_stack.values.insert(0, Some(self.state.clone()));
```

Comment:
can we get around this clone by making `self.state` an option of value and swapping it into the local stack that way we don't need to clone it on every call to aggregate as that would be extremely expensive - but I'm not sure this will work but it might be worth a shot.